core-client/src/main/java/org/glassfish/jersey/client/InvocationInterceptorStages.java

```diff
@@ -0,0 +1,612 @@
+
+        @Override
+        public Locale getLanguage() {
+            return clientResponse.getLanguage();
+        }
+
+        @Override
+        public int getLength() {
+            return clientResponse.getLength();
+        }
+
+        @Override
+        public MediaType getMediaType() {
+            return clientResponse.getMediaType();
+        }
+
+        @Override
+        public Map<String, NewCookie> getCookies() {
+            return clientResponse.getCookies();
+        }
+
+        @Override
+        public EntityTag getEntityTag() {
+            return clientResponse.getEntityTag();
+        }
+
+        @Override
+        public Date getLastModified() {
+            return clientResponse.getLastModified();
+        }
+
+        @Override
+        public URI getLocation() {
+            return clientResponse.getLocation();
+        }
+
+        @Override
+        public Set<Link> getLinks() {
+            return clientResponse.getLinks();
+        }
+
+        @Override
+        public boolean hasLink(String relation) {
+            return clientResponse.hasLink(relation);
+        }
+
+        @Override
+        public Link getLink(String relation) {
+            return clientResponse.getLink(relation);
+        }
+
+        @Override
+        public Link.Builder getLinkBuilder(String relation) {
+            return clientResponse.getLinkBuilder(relation);
+        }
+
+        @Override
+        public boolean hasEntity() {
+            return clientResponse.hasEntity();
+        }
+
+        @Override
+        public InputStream getEntityStream() {
+            return clientResponse.getEntityStream();
+        }
+
+        @Override
+        public void setEntityStream(InputStream input) {
+            clientResponse.setEntityStream(input);
+        }
+    }
+
+    private static ProcessingException createProcessingException(Throwable t) {
+        final ProcessingException processingException = createProcessingException(LocalizationMessages.EXCEPTION_SUPPRESSED());
+        processingException.addSuppressed(t);
+        return processingException;
+    }
+
+    private static ProcessingException createProcessingException(String message) {
+        return new InvocationInterceptorException(message);
+    }
+
+    private static class InvocationInterceptorException extends ProcessingException {
+        private InvocationInterceptorException(String message) {
+            super(message);
+        }
+    }
+
+    private static RuntimeException suppressExceptions(Deque<Throwable> throwables) {
+        if (throwables.size() == 1
+                && RuntimeException.class.isInstance(throwables.getFirst())) {
+            throw (RuntimeException) throwables.getFirst();
+        }
+        final ProcessingException processingException = createProcessingException(LocalizationMessages.EXCEPTION_SUPPRESSED());
+        for (Throwable throwable : throwables) {
+            processingException.addSuppressed(throwable);
+        }
```

Comment:
> the invoker doesn't invoke the interceptor on calling thread

Maybe some other use case would make sense - to implement something similar to `RestClientBuilderListener`. While it could be used by external factories to set the request, it would fit your use case, too. I admit the primary use-case for `PreInvocationInterceptor` is a bit different. 

I monitor requests for such an interface anyway, I try to come up with some PR.

Reply:
Updated in d774b6a